### PR TITLE
feature/security

### PR DESCRIPTION
**Pull request by maiorovvv**

### CODE REVIEW

**Review by Igor-Klein-1988 — APPROVED** [review]

File: pom.xml [hunk 1]
```diff
@@ -16,12 +16,18 @@
...
+        <maven.compiler.encoding>UTF-8</maven.compiler.encoding>
     </properties>
     <dependencies>
         <dependency>
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-data-jpa</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter</artifactId>
```
Comment (Igor-Klein-1988) [inline pom.xml]: Du you use this?

File: src/main/java/de/allcom/config/security/SecurityConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            "/swagger-resources/**",
+            "/configuration/ui",
+            "/configuration/security",
+            "/swagger-ui/**",
+            "/webjars/**",
+            "/swagger-ui.html",
+            "/api/categories/**",
+            "/api/products",
+            "/api/products/**"};
+    // TODO whitelist and security endpoints
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/config/security/SecurityConfiguration.java]: delete this

File: src/main/java/de/allcom/controllers/api/UsersApi.java [hunk 3]
```diff
@@ -1,36 +1,150 @@
...
+            @ApiResponse(responseCode = "200", description = "Successfully retrieved users",
+                    content = @Content(mediaType = "application/json",
+                            array = @ArraySchema(schema = @Schema(implementation = UserAddressResponseDto.class)))),
+            @ApiResponse(responseCode = "401", description = "Unauthorized",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class))),
+            @ApiResponse(responseCode = "403", description = "Forbidden",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class))),
+    })
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/controllers/api/UsersApi.java:56]: What error do you have when  limit =`null`/`-1` or skip =`null`/`-1`?

File: src/main/java/de/allcom/controllers/api/UsersApi.java [hunk 4]
```diff
@@ -1,36 +1,150 @@
...
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class))),
+            @ApiResponse(responseCode = "403", description = "Forbidden",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class))),
+            @ApiResponse(responseCode = "404", description = "User not found",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class)))
+    })
+    UserAddressResponseDto updateUser(@RequestBody UserAddressRegistrationDto request, @PathVariable Long userId);
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/controllers/api/UsersApi.java]: use @Valid
Can you have 400 error's here?

File: src/main/java/de/allcom/controllers/api/UsersApi.java [hunk 5]
```diff
@@ -1,36 +1,150 @@
...
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class))),
+            @ApiResponse(responseCode = "403", description = "Forbidden",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class))),
+            @ApiResponse(responseCode = "404", description = "User not found",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class)))
+    })
+    UserAddressResponseDto foundUserById(@PathVariable Long userId);
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/controllers/api/UsersApi.java]: Can you have 400 error's here?

File: src/main/java/de/allcom/controllers/api/UsersApi.java [hunk 6]
```diff
@@ -1,36 +1,150 @@
...
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class))),
+            @ApiResponse(responseCode = "403", description = "Forbidden",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class))),
+            @ApiResponse(responseCode = "404", description = "User not found",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class)))
+    })
+    UserAddressResponseDto changeStatus(@PathVariable Long userId, @RequestParam String status);
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/controllers/api/UsersApi.java]: What status? 
 Maybe @RequestParam `Boolean` status

Comment (Igor-Klein-1988) [reply src/main/java/de/allcom/controllers/api/UsersApi.java]: Can you have 400 error?

File: src/main/java/de/allcom/controllers/auth/AuthentificationController.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            @ApiResponse(responseCode = "403", description = "Forbidden",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class))),
+            @ApiResponse(responseCode = "409", description = "User with this email already exists",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class)))})
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping("/register")
+    public AuthentificationResponse register(
+            @RequestBody UserAddressRegistrationDto request) {
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/controllers/auth/AuthentificationController.java]: @Valid

File: src/main/java/de/allcom/controllers/auth/AuthentificationController.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            @ApiResponse(responseCode = "403", description = "Forbidden",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class))),
+            @ApiResponse(responseCode = "404", description = "User not found",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class)))
+    })
+    @PostMapping("/login")
+    public AuthentificationResponse login(
+            @RequestBody AuthentificationRequest request) {
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/controllers/auth/AuthentificationController.java]: same

Comment (Igor-Klein-1988) [reply src/main/java/de/allcom/controllers/auth/AuthentificationController.java]: 400 error?

File: src/main/java/de/allcom/controllers/auth/AuthentificationController.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+                            schema = @Schema(implementation = StandardResponseDto.class))),
+            @ApiResponse(responseCode = "403", description = "Forbidden",
+                    content = @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = StandardResponseDto.class))),
+    })
+    @PreAuthorize("hasAnyAuthority('CLIENT', 'ADMIN','STOREKEEPER')")
+    @SecurityRequirement(name = "bearerAuth", scopes = {"admin,user,storekeeper"})
+    @PostMapping("/changePassword")
+    public ResponseEntity<?> changePassword(
+            @RequestBody ChangePasswordRequest request,
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/controllers/auth/AuthentificationController.java]: same

File: src/main/java/de/allcom/controllers/auth/ChangePasswordRequest.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+@Getter
+@Setter
+@Builder
+public class ChangePasswordRequest {
+
+    @Pattern(regexp = "^(?=.*[A-Za-zßäöüÄÖÜ])(?=.*\\d)(?=.*[@#$%^&+=!])(?=\\S+$).{8,}$",
+            message = "Password must be at least 8 characters long and include letters, numbers,"
+                    + " and special characters")
+    @Schema(description = "Current user password", example = "Qwerty007!")
+    private String currentPassword;
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/controllers/auth/ChangePasswordRequest.java]: Do you need `@NotBlank` here?

File: src/main/java/de/allcom/controllers/auth/ChangePasswordRequest.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+    @Pattern(regexp = "^(?=.*[A-Za-zßäöüÄÖÜ])(?=.*\\d)(?=.*[@#$%^&+=!])(?=\\S+$).{8,}$",
+            message = "Password must be at least 8 characters long and include letters, numbers,"
+                    + " and special characters")
+    @Schema(description = "New user password", example = "qweRTY007!")
+    private String newPassword;
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/controllers/auth/ChangePasswordRequest.java]: Do you need `@NotBlank` here?

File: src/main/java/de/allcom/controllers/auth/ChangePasswordRequest.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+    @Pattern(regexp = "^(?=.*[A-Za-zßäöüÄÖÜ])(?=.*\\d)(?=.*[@#$%^&+=!])(?=\\S+$).{8,}$",
+            message = "Password must be at least 8 characters long and include letters, numbers,"
+                    + " and special characters")
+    @Schema(description = "New user confirmation password", example = "qweRTY007!")
+    private String confirmationPassword;
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/controllers/auth/ChangePasswordRequest.java]: Do you need `@NotBlank` here?

File: src/main/java/de/allcom/controllers/auth/AuthentificationResponse.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Schema(description = "User's email", example = "alex-schmidt@mail.com")
+    private String email;
+
+    @Schema(description = "User token",
+            example = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0I")
+    private String token;
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/controllers/auth/AuthentificationResponse.java]: Du you use this?

File: src/main/java/de/allcom/dto/user/UserAddressResponseDto.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
+package de.allcom.dto.user;
+
+import de.allcom.models.Address;
+import de.allcom.models.User;
+import io.swagger.v3.oas.annotations.media.Schema;
+import lombok.Data;
+
+@Data
+public class UserAddressResponseDto {
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/dto/user/UserAddressResponseDto.java]: UserAddressResponseDto -> User`With`AddressResponseDto

File: src/main/java/de/allcom/dto/user/UserAddressResponseDto.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    private String street;
+    @Schema(description = "Company's house number", example = "1")
+    private String houseNumber;
+    @Schema(description = "User's role", example = "ADMIN, CLIENT, STORKEEPER")
+    private String role;
+    @Schema(description = "User checked status", example = "false")
+    private boolean isChecked;
+    @Schema(description = "User blocked status", example = "false")
+    private boolean isBlocked;
+
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/dto/user/UserAddressResponseDto.java]: Make DTO for address

File: src/main/java/de/allcom/dto/user/UserAddressResponseDto.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    private String taxNumber;
+    @Schema(description = "Company's index", example = "10176")
+    private String index;
+    @Schema(description = "Company's city", example = "Berlin")
+    private String city;
+    @Schema(description = "Company's street", example = "Alexanderplatz")
+    private String street;
+    @Schema(description = "Company's house number", example = "1")
+    private String houseNumber;
+    @Schema(description = "User's role", example = "ADMIN, CLIENT, STORKEEPER")
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/dto/user/UserAddressResponseDto.java]: use only one `example `

File: src/main/java/de/allcom/dto/user/UserDto.java [hunk 17]
```diff
@@ -25,12 +25,16 @@ public class UserDto {
...
+    private boolean isBlocked;
+
     public static UserDto from(User user) {
         return UserDto.builder()
                 .id(user.getId())
                 .firstName(user.getFirstName())
                 .lastName(user.getLastName())
                 .email(user.getEmail())
+                .isBlocked(user.isBlocked())
                 .build();
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/dto/user/UserDto.java:49]: Maybe add `isChecked` too

File: src/main/java/de/allcom/exceptions/RestExceptionHandler.java [hunk 18]
```diff
@@ -1,19 +1,21 @@
...
 
 @ControllerAdvice
 public class RestExceptionHandler {
 
-    @ExceptionHandler(value = RestException.class)
+    @ExceptionHandler(RestException.class)
     public ResponseEntity<StandardResponseDto> handleRestException(RestException e) {
         return ResponseEntity
-                .status(e.getStatus())
+                .status(HttpStatus.UNAUTHORIZED)
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/exceptions/RestExceptionHandler.java]: What???????

File: src/main/java/de/allcom/exceptions/SecurityExceptionHandlers.java [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    public static final AuthenticationEntryPoint ENTRY_POINT = (request, response, authException) ->
+            fillResponse(response, HttpStatus.UNAUTHORIZED, "User unauthorized." +  " " + authException.getMessage());
+
+    public static final AccessDeniedHandler ACCESS_DENIED_HANDLER =  (request, response, accessDeniedException) -> {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+        fillResponse(response, HttpStatus.FORBIDDEN, "Access denied for user with email <"
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/exceptions/SecurityExceptionHandlers.java:26]: `email`?

File: src/main/java/de/allcom/models/Address.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+@Getter
+@Setter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+@Entity
+@Table(name = "address")
+public class Address {
+    @Id
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/models/Address.java:24]: add field checks

File: src/main/java/de/allcom/dto/user/UserDto.java (hunk 17 above)
Comment (Igor-Klein-1988) [reply src/main/java/de/allcom/dto/user/UserDto.java:49]: and role too

File: src/main/java/de/allcom/services/UserService.java [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+        return results.map(result -> UserAddressResponseDto.from((User) result[0], (Address) result[1]));
+    }
+
+    public UserAddressResponseDto getUserProfile() {
+        User user = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+        Address address = addressRepository.findByUser(user);
+        return UserAddressResponseDto.from(user, address);
+    }
+
+    public UserAddressResponseDto updateUser(UserAddressRegistrationDto request, Long userId) {
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/services/UserService.java]: add `@Transactional`

File: src/main/java/de/allcom/services/auth/AuthentificationService.java [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+
+    private final TokenRepository tokenRepository;
+
+    private final PasswordEncoder passwordEncoder;
+
+    private final JwtService jwtService;
+
+    private final AuthenticationManager authenticationManager;
+
+    public AuthentificationResponse register(UserAddressRegistrationDto request) {
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/services/auth/AuthentificationService.java]: Maybe add @Transactional

File: src/main/java/de/allcom/services/auth/AuthentificationService.java [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+                    "User with email " + request.getEmail() + " not found!");
+        }
+
+        try {
+            authenticationManager.authenticate(
+                    new UsernamePasswordAuthenticationToken(
+                            request.getEmail(),
+                            request.getPassword()));
+        } catch (Exception e) {
+            throw new RestException(HttpStatus.UNAUTHORIZED, e.getMessage());
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/services/auth/AuthentificationService.java]: Can you repeat this Exception in postman?

File: src/main/java/de/allcom/services/auth/AuthentificationService.java [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+                .user(user)
+                .token(jwtToken)
+                .tokenType(TokenType.BEARER)
+                .revoked(false)
+                .expired(false)
+                .build();
+        tokenRepository.save(token);
+    }
+
+    public ResponseEntity<?> changePassword(ChangePasswordRequest request, Principal connectedUser) {
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/services/auth/AuthentificationService.java]: ResponseEntity<?> ?

File: src/main/resources/db/changelog/changes/202401031338-account-table.xml [hunk 25]
```diff
@@ -22,6 +23,29 @@
             <column name="hash_password" type="varchar(255)">
                 <constraints nullable="false"/>
             </column>
+            <column name="phone_number" type="varchar(13)">
+            </column>
```
Comment (Igor-Klein-1988) [inline src/main/resources/db/changelog/changes/202401031338-account-table.xml:27]: same

File: src/main/java/de/allcom/exceptions/RestExceptionHandler.java (hunk 18 above)
Comment (maiorovvv) [reply src/main/java/de/allcom/exceptions/RestExceptionHandler.java]: It is rebase problem.....

File: src/main/java/de/allcom/exceptions/SecurityExceptionHandlers.java (hunk 19 above)
Comment (maiorovvv) [reply src/main/java/de/allcom/exceptions/SecurityExceptionHandlers.java:26]: Fist of all email it is equals:  User = email.

File: src/main/resources/db/changelog/changes/202401031338-account-table.xml (hunk 25 above)
Comment (maiorovvv) [reply src/main/resources/db/changelog/changes/202401031338-account-table.xml:27]: +490123456789 -> 13 symbols. Ok I make 12symbols.

File: src/main/java/de/allcom/dto/user/UserDto.java [hunk 26]
```diff
@@ -25,12 +28,24 @@ public class UserDto {
...
 
+    @Schema(description = "User's role", example = "ADMIN")
+    private Role role;
+
+    @Schema(description = "User checked status", example = "false")
+    private boolean isChecked;
+
+    @Schema(description = "User blocked status", example = "false")
+    private boolean isBlocked;
+
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/dto/user/UserDto.java:39]: @Schema(description = "Status indicating whether the user is checked", example = "false")
private boolean isChecked;

@Schema(description = "Status indicating whether the user is blocked", example = "false")
private boolean isBlocked;

File: src/main/java/de/allcom/dto/user/UserWithAddressRegistrationDto.java [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Schema(description = "Adddress")
+    @Valid
+    private AddressDto address;
+
+    @Schema(description = "User checked status", example = "false")
+    private boolean isChecked;
+
+    @Schema(description = "User blocked status", example = "false")
+    private boolean isBlocked;
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/dto/user/UserWithAddressRegistrationDto.java:66]: same

File: src/main/java/de/allcom/dto/user/UserWithAddressResponseDto.java [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+    private String position;
+    @Schema(description = "Company's tax's number", example = "123456789")
+    private String taxNumber;
+    @Schema(description = "Address")
+    private AddressDto address;
+    @Schema(description = "User's role", example = "CLIENT")
+    private String role;
+    @Schema(description = "User checked status", example = "false")
+    private boolean isChecked;
+    @Schema(description = "User blocked status", example = "false")
```
Comment (Igor-Klein-1988) [inline src/main/java/de/allcom/dto/user/UserWithAddressResponseDto.java]: same

File: src/main/java/de/allcom/services/auth/AuthentificationService.java (hunk 23 above)
Comment (maiorovvv) [reply src/main/java/de/allcom/services/auth/AuthentificationService.java]: Removed this exception. Error interception is implemented in the class SecurityConfiguration

**Pull request merged**